### PR TITLE
hono

### PR DESCRIPTION
**Pull request by markusahlstrand**
Switch to use hono as router. Think it seems much better than the one I did and would be happy if we don't need to own that code.

### CODE REVIEW

File: src/app.ts [hunk 1]
```diff
@@ -8,25 +9,29 @@ import swaggerUi from "./routes/swagger-ui";
...
-app.use(errorHandler);
-
-app.get("/", async () => {
-  return new Response(
-    JSON.stringify({
-      name: packageJson.name,
-      version: packageJson.version,
-    }),
-  );
+app.use(cors());
```
Comment (markusahlstrand) [inline src/app.ts:28]: Use the cors middleware from hono

File: src/app.ts [hunk 2]
```diff
@@ -82,31 +87,23 @@ app.get("/oauth2-redirect.html", renderOauthRedirectHtml);
 //   return new Response("OK");
 // });
 
-app.get("/static/:file*", serve);
+app.get("/static/:file{.*}", serve);
 
-app.get("/test", async (ctx: Context<Env>) => {
+app.get("/test", async (ctx: Context<{ Bindings: Env }>) => {
```
Comment (markusahlstrand) [inline src/app.ts:92]: The env are passed as Bindings.

File: src/app.ts [hunk 3]
```diff
@@ -82,31 +87,23 @@ app.get("/oauth2-redirect.html", renderOauthRedirectHtml);
...
   const db = getDb(ctx.env);
   const application = await db
     .selectFrom("applications")
     .selectAll()
     .executeTakeFirst();
 
+  const url = new URL(ctx.req.url);
```
Comment (markusahlstrand) [inline src/app.ts:99]: The request object looks a bit different so we need to make a new URL object.

File: src/authentication.ts [hunk 4]
```diff
@@ -158,7 +159,7 @@ async function isValidJwtSignature(
 }
 
 export async function getUser(
-  ctx: Context<Env>,
+  ctx: Context<{ Bindings: Env; Variables: Var }>,
```
Comment (markusahlstrand) [inline src/authentication.ts:162]: Also possible to pass strongly typed vars!

File: src/middlewares/logger.ts [hunk 5]
```diff
@@ -16,18 +22,20 @@ export interface Options {
...
   try {
     const response = await next();
 
-    ctx.event.waitUntil(logResponse(ctx, response!));
+    ctx.executionCtx.waitUntil(logResponse(ctx, response!));
 
     return response;
   } catch (error: any) {
-    ctx.event.waitUntil(logError(ctx, error as Error));
+    ctx.executionCtx.waitUntil(logError(ctx, error as Error));
```
Comment (markusahlstrand) [inline src/middlewares/logger.ts:38]: The waitUntil is now on the executionCtx object. Better choice...

File: src/routes/tsoa/tenants.ts [hunk 6]
```diff
@@ -58,11 +58,11 @@ export class TenantsController extends Controller {
     return data;
   }
 
-  @Get("{id}")
+  @Get("{tenantId}")
```
Comment (markusahlstrand) [inline src/routes/tsoa/tenants.ts:61]: Had to rename this as the hono browser doesn't like having different parameter names in different routes for the same url structure. Probably a good idea..

File: src/services/datadog.ts [hunk 7]
```diff
@@ -111,30 +115,30 @@ async function err(ctx: Context<Env>, err: Error, message?: string) {
...
         " " +
-        request.url +
+        req.url +
         ":" +
         err.message,
     date_access: Date.now(),
     http: {
-      protocol: request.headers.get("X-Forwarded-Proto") || "",
-      host: request.headers.get("host") || "",
+      protocol: req.header("X-Forwarded-Proto") || "",
```
Comment (markusahlstrand) [inline src/services/datadog.ts:125]: This format is nicer for fetching headers..

File: tsoa.mjs [hunk 8]
```diff
@@ -53,7 +53,7 @@ import { readFile, writeFile } from "fs/promises";
     routesDir: "./build",
     controllerPathGlobs: ["src/routes/tsoa/*.ts"],
     authenticationModule: "./src/authentication.ts",
-    middlewareTemplate: "node_modules/tsoa-workers/cloudworker-router.hbs",
+    middlewareTemplate: "node_modules/tsoa-hono/hono-router.hbs",
```
Comment (markusahlstrand) [inline tsoa.mjs:56]: Switch to use the new template for generating the routes.

File: package.json [hunk 9]
```diff
@@ -49,24 +49,24 @@
...
+    "tsoa-hono": "^1.0.3",
     "zod": "^3.22.2"
   },
   "devDependencies": {
     "@cloudflare/workers-types": "^4.20230914.0",
     "@semantic-release/git": "^10.0.1",
     "@types/bcryptjs": "^2.4.4",
     "@types/jest": "^29.5.5",
-    "@types/node": "^20.6.2",
+    "@types/node": "^20.6.3",
```
Comment (danielgent) [inline package.json:69]: this should match the version in the NVM though right?

File: package.json [hunk 10]
```diff
@@ -49,24 +49,24 @@
...
+    "hono": "^3.6.3",
     "kysely": "^0.26.3",
     "kysely-d1": "^0.3.0",
     "kysely-planetscale": "^1.4.0",
     "liquidjs": "^10.9.2",
     "mime": "^3.0.0",
     "trpc-durable-objects": "^1.3.4",
     "tsoa": "^5.1.1",
-    "tsoa-workers": "^1.3.6",
+    "tsoa-hono": "^1.0.3",
```
Comment (danielgent) [inline package.json:61]: I suppose there's one less repo of code you need to maintain now, but there's still this :thinking:

File: src/app.ts [hunk 11]
```diff
@@ -8,25 +9,29 @@ import swaggerUi from "./routes/swagger-ui";
 import rotateKeys from "./routes/rotate-keys";
 import { serve } from "./routes/login";
 import { migrateDown, migrateToLatest } from "./migrate";
-import errorHandler from "./middlewares/errorHandler";
-import corsMiddleware from "./middlewares/cors";
```
Comment (danielgent) [inline src/app.ts:12]: can we delete these two files?

File: package.json (hunk 10 above)
Comment (markusahlstrand) [reply package.json:61]: Yepp, it's just a handlebars template. Wish someone else built it, but.. unfortunately this is still a missing puzzle piece

File: package.json (hunk 9 above)
Comment (markusahlstrand) [reply package.json:69]: Never thought about it but it would make sense.

File: src/app.ts (hunk 11 above)
Comment (danielgent) [reply src/app.ts:12]: By that I mean the `usted` polite tense of `borra estos archivos señor`  :smile: e.g. do it?

This project is already fairly complex. I'd rather nuke files we don't use!

Comment (danielgent) [reply src/app.ts:12]: Ignore that! You already did it. I didn't refresh :upside_down_face:

Comment (danielgent) [reply src/app.ts:12]: Where are we actually configuring the CORS though? e.g. the `tenant-id` header? 

I'm sure *all will be revealed* as I scroll down _but_ we aren't testing this

File: package.json (hunk 9 above)
Comment (danielgent) [reply package.json:69]: Yeah it's quite duplicated isn't it!

In the nextjs projects we also have
```
  "engines": {
    "node": "18.x"
  },
```
in the `package.json`!

File: src/app.ts [hunk 12]
```diff
@@ -82,31 +87,23 @@ app.get("/oauth2-redirect.html", renderOauthRedirectHtml);
...
-
-  return new Response("OK");
-});
-
-app.use(bodyparser);
-
-RegisterRoutes(app);
-
-app.use(app.allowedMethods());
+RegisterRoutes(app as unknown as Hono);
```
Comment (danielgent) [inline src/app.ts:109]: :smiling_face_with_tear:  this `unknown` is a  shame

File: src/middlewares/logger.ts (hunk 5 above)
Comment (danielgent) [reply src/middlewares/logger.ts:38]: Ah, this is what I commented on above

File: src/services/datadog.ts [hunk 13]
```diff
@@ -43,34 +45,32 @@ async function log(
...
       headers,
       status_code: response.status,
-      method: request.method,
-      url_details: request.url,
-      referer: request.headers.get("referer") || "",
-      body: ctx.body,
+      method: req.method,
+      url_details: req.url,
+      referer: req.header("referer") || "",
+      // body: ctx.body,
```
Comment (danielgent) [inline src/services/datadog.ts:58]: :eyes: is this correct?

**Review by danielgent — APPROVED** [review]
LGTM!

File: src/services/datadog.ts (hunk 13 above)
Comment (markusahlstrand) [reply src/services/datadog.ts:58]: Yeah.. I skipped the body as I wasn't sure how to do it. Think the correct thing would be to check the content-type and parse it accordingly. I'm also not 100% sure if there can be issues with consuming the body twice..

File: src/app.ts (hunk 12 above)
Comment (markusahlstrand) [reply src/app.ts:109]: Agree.. I think there is a better solution. I'm considering if we could pass a type to the RegisterRoutes function and make all routes a generic? Soo, something like:
```
RegisterRoutes<{Bindings: Env, Variables: Var>}(app);
```
Would require a patch of the tsoa-library but would be better..

**Pull request merged**